evap/staff/views.py

```diff
@@ -1093,6 +1093,31 @@ def course_copy(request, course_id):
     )
 
 
+@require_POST
+@manager_required
+def create_exam_evaluation(request):
+    evaluation = get_object_from_dict_pk_entry_or_logged_40x(Evaluation, request.POST, "evaluation_id")
+    if evaluation.is_single_result:
+        raise SuspiciousOperation("Creating an exam evaluation for a single result evaluation is not allowed")
+
+    if evaluation.has_exam:
+        raise SuspiciousOperation("An exam evaluation already exists for this course")
+    try:
+        exam_date = request.POST.get("exam_date")
+        exam_date = datetime.strptime(exam_date, "%Y-%m-%d").date()
+    except TypeError:
+        return HttpResponseBadRequest("Exam date missing or invalid.")
+
+    if exam_date < evaluation.vote_start_datetime.date() + timedelta(days=1):
+        raise SuspiciousOperation(
+            "The end date of the main evaluation would be before its start date. No exam evaluation was created."
+        )
+
+    evaluation.create_exam_evaluation(exam_date)
```

Comment:
I do not have a strong opinion on where the check should be, but kinda like it here as this case should never happen from user interaction and it is really not an interal server error.

Reply:
Yeah looking at it today again, I also think that internal server error would be wrong. One argument for checking it inside `create_exam_evaluation` is that we explicitly made everything inside `create_exam_evaluation` in a `transaction.atomic`, but the check here is outside of that. I think the correct-est way would be something like
```python
# Model code:
def create_exam_evaluation(self, exam_date):
    if exam_date < self.earliest_possible_exam_date:
        raise ValueError("...")
    ...

# View code:
def create_exam_evaluation(request):
    ...
    try:
        exam_evaluation = evaluation.create_exam_evaluation(...)
    except ValueError as e:
        raise SuspiciousOperation from e
```
What do you think? Also, thoughts @richardebeling @Kakadus ?

Reply:
View-Validation here is fine for me. The field has a [constraint](https://github.com/e-valuation/EvaP/blob/6d7970e23b4c51eea2063b8ceb681ec9398edc2f/evap/evaluation/models.py#L489) that ensures we can not store "wrong" data. Attempting to do that would correctly cause an internal server error. A complying view checks/handles this, and instead returns a client error.

I wouldn't do `raise ValueError` in `create_exam_evaluation`, that would be a third check, seems overly complicated to me. The view could in theory try to catch a constraint validation error, but I think I like the current approach most.

Reply:
I don't mean to add a third check, I would just move all checks into the transaction and close to the creation. Main motivation for me is "parse, don't validate". It would then look like this:
```python
class Evaluation:
    # ...

    @transaction.atomic
    def create_exam_evaluation(self, exam_date: date):
        if self.is_single_result:
            raise ValueError("Creating an exam evaluation for a single result evaluation is not allowed")
        if self.has_exam_evaluation:
            raise ValueError("An exam evaluation already exists for this course")
        if exam_date < self.earliest_possible_exam_date:
            raise ValueError("The end date of the main evaluation would be before its start date.")

        self.weight = 9
        self.vote_end_date = exam_date - timedelta(days=1)
        self.save()
        exam_evaluation = Evaluation(
            course=self.course,
            name_de="Klausur",
            name_en="Exam",
            weight=1,
            is_rewarded=False,
            vote_start_datetime=datetime.combine(exam_date + timedelta(days=1), time(8, 0)),
            vote_end_date=exam_date + timedelta(days=3),
        )
        exam_evaluation.save()

        exam_evaluation.participants.set(self.participants.all())
        for contribution in self.contributions.exclude(contributor=None):
            exam_evaluation.contributions.create(contributor=contribution.contributor)
        exam_evaluation.general_contribution.questionnaires.set(settings.EXAM_QUESTIONNAIRE_IDS)
        exam_evaluation.save()

# views.py
@require_POST
@manager_required
def create_exam_evaluation(request: HttpRequest) -> HttpResponse:
    evaluation = get_object_from_dict_pk_entry_or_logged_40x(Evaluation, request.POST, "evaluation_id")
    exam_date_string = request.POST.get("exam_date", "")

    try:
        exam_date = datetime.strptime(exam_date_string, "%Y-%m-%d").date()
        evaluation.create_exam_evaluation(exam_date)
    except ValueError as e:
        raise SuspiciousOperation("Failed to create exam evaluation") from e

    messages.success(request, _("Successfully created exam evaluation."))
    return HttpResponse()  # 200 OK
```
@richardebeling Which do you prefer? I like the above, but if you are still with the current one, I can accept also accept that

Reply:
@niklasmohrin is this resolved for you?

Reply:
Yeah, you can close it